### PR TITLE
Allow configuring regions from tfvars in FAST networking stages

### PR DESCRIPTION
**Pull request by ludoo**
This has bugged me for a long time: if I want to change regions in our networking stages, I need to edit the code in a lot of places, including module names and references (`module.landing-to-onprem-ew1-vpn` etc.).

The `region_trigrams` variable is also pretty evil, and would wreak havoc with subnets added via factory for regions not accounted for in variables.

So this PR proposes a way to abstract the region definition in variables via `primary` and `secondary` regions, and generating region abbreviations on the fly for all regions present in subnets.

It only does so for the peering stage to validate the concept, it will of course be fairly easy to roll this out to the other network stages. I ran the stage against the current one and plan did not show any difference apart from the change in module names via `moved` blocks.

As an aside, I also think that L7 ILB subnets and PSA should come via factory, to minimize the variable surface and offer (almost) full configuration via delcarative files. I plan to tackle this once the regions thing is sorted.

### CODE REVIEW

File: fast/stages/2-networking-a-peering/main.tf [hunk 1]
```diff
@@ -17,13 +17,20 @@
...
+    values(module.prod-spoke-vpc.subnet_regions),
+  ))
   custom_roles = coalesce(var.custom_roles, {})
-  l7ilb_subnets = {
-    for env, v in var.l7ilb_subnets : env => [
-      for s in v : merge(s, {
-        active = true
-        name   = "${env}-l7ilb-${s.region}"
-    })]
+  region_shortnames = {
```
Comment (sruffilli) [inline fast/stages/2-networking-a-peering/main.tf:28]: This would misbehave in case of regions with a double digit (e.g. `europe-westXX`)

File: fast/stages/2-networking-a-peering/variables.tf [hunk 2]
```diff
@@ -167,12 +173,15 @@ variable "psa_ranges" {
   # }
 }
 
-variable "region_trigram" {
-  description = "Short names for GCP regions."
-  type        = map(string)
+variable "regions" {
```
Comment (sruffilli) [inline fast/stages/2-networking-a-peering/variables.tf:176]: Long term I'd rather have regions controlled in a single location, possibly defined in stage 0.

**Review by sruffilli — APPROVED** [review]
LGTM pending a small fix on trigram definition.

File: fast/stages/2-networking-a-peering/variables.tf (hunk 2 above)
Comment (ludoo) [reply fast/stages/2-networking-a-peering/variables.tf:176]: maybe... I'm not convinced regions are org-wide vs network-stage specific, let's discuss

Comment (ludoo) [reply fast/stages/2-networking-a-peering/variables.tf:176]: well... it would make org policies simpler so I see your point

Comment (ludoo) [reply fast/stages/2-networking-a-peering/variables.tf:176]: let's move this to an issue so we can track it

File: fast/stages/2-networking-a-peering/main.tf (hunk 1 above)
Comment (juliocc) [reply fast/stages/2-networking-a-peering/main.tf:28]: maybe
```hcl
region_shortnames = {
  for v in local._regions : 
  v => join("", regex("([a-z])[a-z]+-([a-z])[a-z]+(\\d+)", "europe-west1"))
}
```

But that will still not work with something like asia-south1 and asia-southest1

File: fast/stages/2-networking-a-peering/variables.tf [hunk 3]
```diff
@@ -167,12 +173,15 @@ variable "psa_ranges" {
...
+  description = "Region definitions."
+  type = object({
+    primary   = string
+    secondary = string
```
Comment (juliocc) [inline fast/stages/2-networking-a-peering/variables.tf:180]: would it make sense to support more than one secondary?

File: fast/stages/2-networking-a-peering/main.tf (hunk 1 above)
Comment (ludoo) [reply fast/stages/2-networking-a-peering/main.tf:28]: this is what the current implementation spits out, I will update as per Simo's point and share the updated results

```hcl
tmp = {
  "asia-east1" = "ae1"
  "asia-east2" = "ae2"
  "asia-northeast1" = "an1"
  "asia-northeast2" = "an2"
  "asia-northeast3" = "an3"
  "asia-south1" = "as1"
  "asia-south2" = "as2"
  "asia-southeast1" = "as1"
  "asia-southeast2" = "as2"
  "australia-southeast1" = "as1"
  "australia-southeast2" = "as2"
  "europe-central2" = "ec2"
  "europe-north1" = "en1"
  "europe-southwest1" = "es1"
  "europe-west1" = "ew1"
  "europe-west2" = "ew2"
  "europe-west3" = "ew3"
  "europe-west4" = "ew4"
  "europe-west6" = "ew6"
  "europe-west8" = "ew8"
  "europe-west9" = "ew9"
  "me-west1" = "mw1"
  "northamerica-northeast1" = "nn1"
  "northamerica-northeast2" = "nn2"
  "southamerica-east1" = "se1"
  "southamerica-west1" = "sw1"
  "us-central1" = "uc1"
  "us-east1" = "ue1"
  "us-east4" = "ue4"
  "us-east5" = "ue5"
  "us-south1" = "us1"
  "us-west1" = "uw1"
  "us-west2" = "uw2"
  "us-west3" = "uw3"
  "us-west4" = "uw4"
}
```

File: fast/stages/2-networking-a-peering/main.tf [hunk 4]
```diff
@@ -17,13 +17,20 @@
...
+    for k, v in local._region_tokens : k => join("", [
+      substr(v.0, 0, 1),
+      substr(v.1, 0, 1),
+      substr(strrev(v.1), 0, 1),
+    ])
```
Comment (juliocc) [inline fast/stages/2-networking-a-peering/main.tf:33]: I really hate this. Lots of magic numbers and hard to follow. See proposal above

File: fast/stages/2-networking-a-peering/main.tf (hunk 1 above)
Comment (ludoo) [reply fast/stages/2-networking-a-peering/main.tf:28]: so, slight tweak and I think it works

```hcl
  region_shortnames = {
    for k, v in local._regions : k => join("", [
      substr(v.0, 0, 1),
      substr(v.1, 0, 1),
      regex("[0-9]+$", v.1),
    ])
  }
  _regions = { for v in local.regions : v => split("-", v) }
  ```
  
result

```hcl
tmp = {
  "asia-east1" = "ae1"
  "asia-east2" = "ae2"
  "asia-northeast1" = "an1"
  "asia-northeast2" = "an2"
  "asia-northeast3" = "an3"
  "asia-south1" = "as1"
  "asia-south2" = "as2"
  "asia-southeast1" = "as1"
  "asia-southeast2" = "as2"
  "australia-southeast1" = "as1"
  "australia-southeast2" = "as2"
  "europe-central2" = "ec2"
  "europe-north1" = "en1"
  "europe-southwest1" = "es1"
  "europe-west1" = "ew1"
  "europe-west12" = "ew12"
  "europe-west2" = "ew2"
  "europe-west3" = "ew3"
  "europe-west4" = "ew4"
  "europe-west6" = "ew6"
  "europe-west8" = "ew8"
  "europe-west9" = "ew9"
  "me-west1" = "mw1"
  "northamerica-northeast1" = "nn1"
  "northamerica-northeast2" = "nn2"
  "southamerica-east1" = "se1"
  "southamerica-west1" = "sw1"
  "us-central1" = "uc1"
  "us-east1" = "ue1"
  "us-east4" = "ue4"
  "us-east5" = "ue5"
  "us-south1" = "us1"
  "us-west1" = "uw1"
  "us-west2" = "uw2"
  "us-west3" = "uw3"
  "us-west4" = "uw4"
}
```

Comment (ludoo) [reply fast/stages/2-networking-a-peering/main.tf:28]: ouch, asia southeast

File: fast/stages/2-networking-a-peering/variables.tf (hunk 3 above)
Comment (ludoo) [reply fast/stages/2-networking-a-peering/variables.tf:180]: easy to add when/if needed

File: fast/stages/2-networking-a-peering/main.tf (hunk 4 above)
Comment (ludoo) [reply fast/stages/2-networking-a-peering/main.tf:33]: I really hate regexps lol

**Comment by ludoo** [comment]
needs actual testing for vpn/nva/separated

**Review by juliocc — APPROVED** [review]
